pylops/optimization/solver.py

```diff
@@ -2,10 +2,186 @@
+        """Setup solver
+
+        Parameters
+        ----------
+        x0 : :obj:`np.ndarray`, optional
+            Initial guess
+        niter : :obj:`int`, optional
+            Number of iterations
+        tol : :obj:`float`, optional
+            Tolerance on residual norm
+        show : :obj:`bool`, optional
+            Display setup log
+
+        """
+        self.tol = tol
+        self.niter = niter
+
+        # initialize solver
+        self.x0 = x0
+        if self.x0 is None:
+            self.x = self.ncp.zeros(self.Op.shape[1], dtype=self.y.dtype)
+            self.r = self.y.copy()
+        else:
+            self.x = self.x0.copy()
+            self.r = self.y - self.Op.matvec(self.x)
+        self.c = self.r.copy()
+        self.kold = self.ncp.abs(self.r.dot(self.r.conj()))
+
+        # create variables to track the residual norm and iterations
+        self.cost = []
+        self.cost.append(np.sqrt(self.kold))
+        self.iiter = 0
+
+        # print setup
+        if show:
+            self._print_setup()
+
+    def step(self, show=False):
+        """Run one step of solver
+
+        Parameters
+        ----------
+        show : :obj:`bool`, optional
+            Display iteration log
+
+        """
+        Opc = self.Op.matvec(self.c)
+        cOpc = self.ncp.abs(self.c.dot(Opc.conj()))
+        a = self.kold / cOpc
+        self.x += a * self.c
+        self.r -= a * Opc
+        k = self.ncp.abs(self.r.dot(self.r.conj()))
+        b = k / self.kold
+        self.c = self.r + b * self.c
+        self.kold = k
+        self.iiter += 1
+        self.cost.append(np.sqrt(self.kold))
+        if show:
+            self._print_step()
+
+    def finalize(self, show=False):
+        """Finalize solver
+
+        Parameters
+        ----------
+        show : :obj:`bool`, optional
+            Display finalize log
+
+        """
+        if show:
+            self._print_finalize()
+        self.cost = np.array(self.cost)
+
+    def solve(self, x0=None, niter=10, tol=1e-4, show=False):
+        """Run entire solver
+
+        Parameters
+        ----------
+        x0 : :obj:`np.ndarray`, optional
+            Initial guess
+        niter : :obj:`int`, optional
+            Number of iterations
+        tol : :obj:`float`, optional
+            Tolerance on residual norm
+        show : :obj:`bool`, optional
+            Display setup log
+
+        Returns
+        -------
+        x : :obj:`np.ndarray`
+            Estimated model of size :math:`[N \times 1]`
+        iit : :obj:`int`
+            Number of executed iterations
+        cost : :obj:`numpy.ndarray`, optional
+            History of the L2 norm of the residual
+
+        """
+        self.setup(x0=x0, niter=niter, tol=tol, show=show)
+        while self.iiter < niter and self.kold > self.tol:
+            show = (
+                True
+                if self.iiter < 10 or niter - self.iiter < 10 or self.iiter % 10 == 0

```

Comment:
Consider not hard-coding 10.

Reply:
> I would also like initial guess as optional
> 
> ```
> out=cg.solve(y, x0) 
> ```
> 

It is in now :)

Reply:
And good point about not hard-coding 10... this is spicy behavior, reason why I followed it in the first place but we can make it more flexible for sure

Reply:
Have a look at the new parameter ``itershow``